### PR TITLE
Big int cohort id fix

### LINKED ISSUE #162: Use of big integer outcome, target and comparator ids causes `checkmate` failures
There are also issues in the results data model as this specifies only int values.

PR incoming that uses `checkmate::assertNumeric` and `id %% 1 = 0` as proxy for large int values.

### PR DESCRIPTION
**Pull request by azimov**
Resolves #162

### CODE REVIEW

File: R/Analyses.R [hunk 1]
```diff
@@ -210,19 +210,19 @@ createOutcome <- function(outcomeId,
...
+  checkmate::assertNumeric(outcomeId, add = errorMessages)
+  checkmate::assertTRUE(all(outcomeId %% 1 == 0), add = errorMessages)
   checkmate::assertLogical(outcomeOfInterest, add = errorMessages)
   checkmate::assertNumeric(trueEffectSize, len = 1, null.ok = TRUE, add = errorMessages)
   checkmate::assertInt(riskWindowStart, null.ok = TRUE, add = errorMessages)
   checkmate::assertInt(riskWindowEnd, null.ok = TRUE, add = errorMessages)
   checkmate::reportAssertions(collection = errorMessages)
   if (!is.null(startAnchor) && !grepl("start$|end$", startAnchor, ignore.case = TRUE)) {
-    stop("startAnchor should have value 'cohort start' or 'cohort end'")
+    stop("startAnchor should have value \'cohort start\' or \'cohort end\'")
```
Comment (schuemie) [inline R/Analyses.R:221]: Could you help me understand why you're escaping single quotes? It is not needed when the string uses double quotes

File: R/RunAnalyses.R [hunk 2]
```diff
@@ -292,13 +292,14 @@ runCmAnalyses <- function(connectionDetails,
...
-  subset <- subset[!file.exists(file.path(outputFolder, subset$cohortMethodDataFile)), ]
+  subset <- referenceTable[!duplicated(referenceTable$cohortMethodDataFile),]
+  subset <- subset[subset$cohortMethodDataFile != "",]
+  subset <- subset[!file.exists(file.path(outputFolder, subset$cohortMethodDataFile)),]
   if (nrow(subset) != 0) {
     message("*** Creating cohortMethodData objects ***")
+
     createCmDataTask <- function(i) {
-      refRow <- subset[i, ]
+      refRow <- subset[i,]
```
Comment (schuemie) [inline R/RunAnalyses.R:302]: Per our code guidelines, a comma should be followed by a space. Why remove it here?

File: R/RunAnalyses.R [hunk 3]
```diff
@@ -493,24 +500,30 @@ runCmAnalyses <- function(connectionDetails,
...
         list(analysisId = refRow$analysisId)
       )[[1]]
-      task <- list(
-        psFile = file.path(outputFolder, refRow$psFile),
-        args = analysisRow,
-        strataFile = file.path(outputFolder, refRow$strataFile)
-      )
-      return(task)
+
+      if (file.exists(file.path(outputFolder, refRow$psFile))) {
```
Comment (schuemie) [inline R/RunAnalyses.R:516]: Why the condition? Is there something wrong with the code for generating the reference table that will specify strataFiles to be created without there being psFiles? Maybe best to fix it there?

File: R/RunAnalyses.R [hunk 4]
```diff
@@ -825,19 +848,24 @@ addPsToStudyPopForSubset <- function(subset, outputFolder) {
   ps <- readRDS(file.path(outputFolder, subset$sharedPsFile[1]))
 
   addToStudyPop <- function(i) {
-    refRow <- subset[i, ]
+    refRow <- subset[i,]
     studyPop <- readRDS(file.path(outputFolder, refRow$studyPopFile))
     studyPop <- addPsToStudyPopulation(studyPop, ps)
-    saveRDS(studyPop, file.path(outputFolder, refRow$psFile))
+    tryCatch({
```
Comment (schuemie) [inline R/RunAnalyses.R:854]: Is the error message thrown by saveRDS not informative enough? If so, maybe we should have a generic wrapper for saveRDS in CohortMethod that we always use?

File: R/RunAnalyses.R [hunk 5]
```diff
@@ -892,100 +920,118 @@ applyTrimMatchStratify <- function(ps, arguments) {
 }
 
 doTrimMatchStratify <- function(params) {
-  ps <- getPs(params$psFile)
-  ParallelLogger::logDebug(sprintf("Performing matching etc., using %s",
-                                   params$psFile))
-  ps <- applyTrimMatchStratify(ps, params$args)
-  saveRDS(ps, params$strataFile)
+  tryCatch({
```
Comment (schuemie) [inline R/RunAnalyses.R:923]: Yikes, this tryCatch block covers multiple function calls, but the error message will be about the saving of the stataFile.

File: R/RunAnalyses.R [hunk 6]
```diff
@@ -892,100 +920,118 @@ applyTrimMatchStratify <- function(ps, arguments) {
...
-    }
-    covariatesToInclude <- unique(c(covariatesToInclude, params$args$interactionCovariateIds))
-    if (length(covariatesToInclude) != 0) {
-      covariates <- covariates %>%
-        filter(.data$covariateId %in% covariatesToInclude)
-    }
-    if (length(covariatesToExclude) != 0) {
-      covariates <- covariates %>%
-        filter(!.data$covariateId %in% covariatesToExclude)
+  tryCatch({
```
Comment (schuemie) [inline R/RunAnalyses.R:938]: Double yikes! What is the point in wrapping such a large block of code in a try catch?

File: R/RunAnalyses.R [hunk 7]
```diff
@@ -892,100 +920,118 @@ applyTrimMatchStratify <- function(ps, arguments) {
...
+      covariates = covariates,
+      covariateRef = cohortMethodData$covariateRef,
+      analysisRef = cohortMethodData$analysisRef
+    )
+    attr(filteredCohortMethodData, "metaData") <- attr(cohortMethodData, "metaData")
+    class(filteredCohortMethodData) <- "CohortMethodData"
+    attr(class(filteredCohortMethodData), "package") <- "CohortMethod"
+    saveCohortMethodData(filteredCohortMethodData, params$prefilteredCovariatesFile)
+  }, error = function(err) {
+    ParallelLogger::logError(err)
```
Comment (schuemie) [inline R/RunAnalyses.R:971]: This code has no effect. Errors are automatically caught by ParalllelLogger. Just rethrowing the error will achieve the same thing

**Comment by schuemie** [comment]
I'm sorry, I can't accept this PR. It contains a lot of changes that have nothing to do with the issue at hand. 

I'm open to rethinking how we deal with errors, but that should be a separate issue and a separate PR (but only after we agree on the approach).

Also, please adhere to the style guide (as also implemented by the styler package).

**Comment by azimov** [comment]
My apologies, I did a diff with the wrong branch and included the code I used to find issues. This now has been removed.

From: Martijn Schuemie ***@***.***>
Date: Sunday, April 7, 2024 at 10:20 PM
To: OHDSI/CohortMethod ***@***.***>
Cc: Gilbert, James [JANUS] ***@***.***>, Author ***@***.***>
Subject: [EXTERNAL] Re: [OHDSI/CohortMethod] Big int cohort id fix (PR #163)

I'm sorry, I can't accept this PR. It contains a lot of changes that have nothing to do with the issue at hand.

I'm open to rethinking how we deal with errors, but that should be a separate issue and a separate PR (but only after we agree on the approach).

Also, please adhere to the style guide (as also implemented by the styler package).

**Comment by schuemie** [comment]
Ah, the problem is that SQLite does not allow changing the column type of an existing table. The good news is that SQLite INT is dynamic, and will contain BIGINT values. 

Why do we have cohort IDs > 2^31?

**Comment by azimov** [comment]
> Why do we have cohort IDs > 2^31?

These are generally cohorts made from templates or subsets where we do something like `concept_id * 1000 as cohort_definition_id`, this happened when running this on the reward cohorts

**Comment by schuemie** [comment]
Ok, we'll probably need fixes throughout HADES then. I always assumed, since we're assigning cohort IDs ourselves, we would not exceed 2^31.

I think I fixed the unit tests now, by fixing the SqlRender translation of `ALTER TABLE ALTER COLUMN`. 

I also removed the PostgreSQL-specific version of the migration script. We really shouldn't have platform-specific code in HADES packages other than SqlRender and DatabaseConnector.

**Pull request merged**